.-.

#Máster Universitario en Inteligencia Artificial: 02MIAR: Matemáticas para la IA
####Ejercicios evaluables: Cálculo Multivariable: El Descenso de Gradiente
#**Grupo 17** Paula Mejias, Julian Samblas, Raul Contreras O.
###Link:  https://colab.research.google.com/drive/1B1A5fFFymsXTbFtl9tOOtYWY_btsSFrD#scrollTo=L_VxUQPJlUog
###Github:  

---



Álgebra Lineal y Optimización Computacional
El álgebra lineal constituye el pilar fundamental sobre el cual se edifican diversas técnicas de aprendizaje automático (machine learning). Sus herramientas son indispensables para implementar procesos críticos, entre los cuales destacan:

Propagación hacia adelante (Forward propagation): La base de la computación en redes neuronales.bb

Optimización multivariable: Identificación de mínimos locales, proceso esencial en la retropropagación (backpropagation).

Reducción de dimensionalidad: El uso de técnicas como el Análisis de Componentes Principales (PCA) para simplificar estructuras de datos complejas.

Dada la escala masiva de los datasets modernos, es crucial emplear algoritmos altamente optimizados que minimicen el coste computacional. A continuación, se presentan diversas alternativas para abordar operaciones de álgebra lineal, evaluando el impacto de cada variante en términos de eficiencia. Para este estudio, nos centraremos específicamente en el cálculo del determinante de una matriz.

   **a) Implementa una función, determinante recursiva, que obtenga el determinante de una matriz cuadrada utilizando la definicion recursiva de Laplace** [1 punto]

#Las Matrices de Ejemplo

   1 Matriz Normal ($A$): Sin ceros en la diagonal principal.
   
   $$A = \begin{bmatrix} 5 & 2 & 9 \\ 1 & 3 & 4 \\ 8 & 6 & 2 \end{bmatrix}$$


   2 Matriz con Diagonal Cero ($B$): Útil para demostrar la necesidad del pivoteo  obligatorio.
   
   $$B = \begin{bmatrix} 0 & 2 & 9 \\ 1 & 0 & 4 \\ 8 & 6 & 0 \end{bmatrix}$$


---




**a) Método de Laplace (Expansión por Cofactores)**
   
   Laplace se basa en reducir el problema: $\det(A) = \sum_{j=0}^{n-1} (-1)^{0+j} \cdot a_{0j} \cdot \det(M_{0j})$
   
  -   $a_{0j}$: Elemento en la fila 0, columna $j$.
  
  -   $(-1)^{0+j}$: El "signo de tablero de ajedrez" (alterna entre + y -).
  
  -   $M_{0j}$: La submatriz resultante de eliminar la fila 0 y la columna $j$.





---



**Ejemplo con Matriz $A$:**

Tomamos la primera fila ($5, 2, 9$):


a)     $+5 \cdot \det\begin{bmatrix} 3 & 4 \\ 6 & 2 \end{bmatrix} = 5 \cdot (6-24) = 5 \cdot (-18) = -90$




b)     $-2 \cdot \det\begin{bmatrix} 1 & 4 \\ 8 & 2 \end{bmatrix} = -2 \cdot (2-32) = -2 \cdot (-30) = +60$


c)     $+9 \cdot \det\begin{bmatrix} 1 & 3 \\ 8 & 6 \end{bmatrix} = 9 \cdot (6-24) = 9 \cdot (-18) = -162$


Resultado: $\det(A) = -90 + 60 - 162 = -192$.




---




**Ejemplo con Matriz  B :**

Desarrollo de Laplace Paso a Paso (Fila 1: $0, 2, 9$)
Usamos la primera fila como cofactores. La fórmula de expansión nos da:

$$\det(B) = +(\text{Cofactor}_1 \cdot \text{Submatriz}_1) - (\text{Cofactor}_2 \cdot \text{Submatriz}_2) + (\text{Cofactor}_3 \cdot \text{Submatriz}_3)$$
  
1. Primer término (Elemento $0$):$$+ (0) \cdot \det \begin{bmatrix} 0 & 4 \\ 6 & 0 \end{bmatrix} = 0$$


2. Segundo término (Elemento $2$, con signo negativo):$$- (2) \cdot \det \begin{bmatrix} 1 & 4 \\ 8 & 0 \end{bmatrix}$$


 Calculamos el determinante de la submatriz: $(1 \cdot 0) - (4 \cdot 8) = 0 - 32 = -32$.
Entonces: $-2 \cdot (-32) = \mathbf{+64}$.

3. Tercer término (Elemento $9$, con signo positivo):$$+ (9) \cdot \det \begin{bmatrix} 1 & 0 \\ 8 & 6 \end{bmatrix}$$Calculamos el determinante de la submatriz: $(1 \cdot 6) - (0 \cdot 8) = 6 - 0 = 6$.Entonces: $9 \cdot 6 = \mathbf{+54}$.


Resultado final:$$\det(B) = 0 + 64 + 54 = 118$$

---




Slicing: Extraemos la submatriz (el Menor).

Ejemplo con tu matriz 3x3:Si $A$ es:


$$\begin{bmatrix}
\mathbf{a_{00}} & \mathbf{a_{01}} & \mathbf{a_{02}} \\
\mathbf{a_{10}} & a_{11} & a_{12} \\
\mathbf{a_{20}} & a_{21} & a_{22}
\end{bmatrix}$$


Al aplicar matriz[1:, 1:], le estás diciendo a Python: "Ignora todo lo que está en negrita y quédate solo con esto":


$$\begin{bmatrix}
a_{11} & a_{12} \\
a_{21} & a_{22}
\end{bmatrix}$$


Interpretando tu código: det_sub1 = sub1[0, 0] * sub1[1, 1] - sub1[0, 1] * sub1[1, 0]


Esta línea es el cálculo del determinante de una matriz 2x2. Si tu submatriz sub1 se ve así:


$$\text{sub1} = \begin{bmatrix}
\text{sub1}[0,0] & \text{sub1}[0,1] \\
\text{sub1}[1,0] & \text{sub1}[1,1]
\end{bmatrix}$$

In [23]:
#-------------------------------------------------------------------------------
# Librerias
#-------------------------------------------------------------------------------

import numpy as  np
import time
# NumPy redondea  2 decimales en los print
np.set_printoptions(precision=2, suppress=True)
# lru_cache Permite guardar los resultados de una llamada de funcion recursiva costosa en
# un cache de memoria cuando la misma informacion ocurre nuevamente
from functools import lru_cache
# Permite guardar los resultados de una llamada de función recursiva costosa en
# un caché de memoria cuando la misma información ocurre nuevamente.
# Esto optimiza Laplace de O(n!) a una complejidad mucho más manejable.

In [24]:
#-------------------------------------------------------------------------------
# Metodo de Lapace
#-------------------------------------------------------------------------------

def determinante_recursivo_laplace(matriz, nivel=0, v_print=0):
    """
    Calcula el determinante de forma recursiva (Laplace)
    """
    # Identación visual para el reporte jerárquico
    indent = "  " * nivel
    n = matriz.shape[0]
    # Propiedad de los arreglos de NumPy,devuelve una tupla que indica las dimensiones de la matriz.
    # Por ejemplo,  una matriz de 3 x 3 , shape devuelve (3, 3).[0]:
    # AEl 0, le indica a Python que tome el primer valor de esa tupla. Es decir, nos da el número de filas de la matriz.
    if v_print == 1:
        print(f"\n{indent}=== Tamaño de la matriz (Matriz {n}x{n}) ===")


    # 1. CASO BASE: Matriz 2x2 se detiene la llamada recursiva
    if n == 2:
        return matriz[0, 0] * matriz[1, 1] - matriz[0, 1] * matriz[1, 0]
          # Sub_Matriz =  [3, 4]
          #               [6, 2]
          # = 3 x 2 - 4 x 6
    det = 0
    if v_print == 1:
        print(f"\n{indent}  == Iniciando expansión nivel {nivel} (Matriz {n}x{n}) ===")

    # 2. CASO RECURSIVO: Expandir por la primera fila
    for j in range(n):
        cofactor = matriz[0, j]
           # Término 1: Cofactor B[0,0] = 0
           # Matriz =   [5, 2, 9]
           #            [1, 3, 4]
           #            [8, 6, 2]
           # los cofactores segun valor de J son 5, 2, 9

        # Slicing dinámico: Elimina fila 0 y columna j
        submatriz = np.delete(np.delete(matriz, 0, axis=0), j, axis=1)
        # Paso 1 (Lo de adentro): np.delete(matriz, 0, axis=0):Toma la matriz,
        # busca el índice 0 (la primera fila),
        # y la elimína (axis=0 significa filas)".
        # Paso 2 (Lo de afuera): np.delete(..., j, axis=1), Toma lo que resultó del paso anterior y,
        # a eso, quítale la columna j (axis=1 significa columnas).
        # axis=0: Es el eje vertical. Cuando  se usa, afecta las filas (cortando horizontalmente).
        # axis=1: Es el eje horizontal. Cuando se  usa, afectas las columnas (cortando verticalmente).
        # Delete crea un nuevo arreglo eliminando las filas y columnas indicadas.
        # Sub_Matriz =  [3, 4]
        #               [6, 2]
        # Ejemplo
        # temporal = np.delete(matriz, 0, axis=0)    # Elimina fila
        # submatriz = np.delete(temporal, j, axis=1) # Elimina columna

        # LLAMADA RECURSIVA: La función se llama a sí misma
        det_sub = determinante_recursivo_laplace(submatriz, nivel + 1)

        # Aplicamos la fórmula: (-1)^j * cofactor * Det(submatriz)
        term = ((-1) ** j) * cofactor * det_sub
        det += term

        # 3. PASO A PASO: Impresión de Sub_Matrices
        if v_print == 1:
           print(f"{indent}Término {j+1} (Cofactor {cofactor:.0f}):")
           print(f"{indent}  Submatriz:")

        # BUCLE DINÁMICO: Esto evita el error IndexError
        for fila in submatriz:
            if v_print == 1:
               print(f"{indent}  {fila}")


        signo = "+" if j % 2 == 0 else "-" # Parte de la formula matematica que controla el signo
        # ejemplo En una matriz de 3x3, los índices de las columnas son 0, 1 y 2:
        # Columna j=0: Par   (0) (+)
        # Columna j=1: Impar (1) (-)
        # Columna j=2: Par   (2) (+)
        # Asigna a la variable signo el valor de '+' si el resto de dividir j entre 2 es igual a 0,
        # de lo contrario, asígnale '-' }
        # j % 2 == 0? Es la forma matemática de saber si un número es par.
        #En la expansión de Laplace, los términos alternan de signo según la posición.
        #Esto hace que si estás en la columna 0, 2, 4 (pares), el signo sea positivo,
        # y si estás en la columna 1, 3, 5 (impares), el signo sea negativo.
        #Este el mecanismo que implementa la regla de los signos de los cofactores

        if v_print == 1:
           print(f"{indent}  Operación: {signo} {cofactor:.0f} * ({det_sub:.2f})")
           print(f"{indent}  Resultado Término {j+1} = {term:.4f}")

    return det




**b) Si A es una matriz cuadrada n×n y triangular (superior o inferior, es decir,
con entradas nulas por debajo o por encima de la diagonal, respectivamente), ¿existe alguna forma de calcular de forma directa y sencilla su determinante? Justifıquese la respuesta.[0.5 puntos]**

Para  calcular el determinante de una matriz triangular  $A$ ya sea superior o inferior, su determinante es igual al producto de los  elementos de su diagonal principal.  

$$\det(A) = \prod_{i=1}^{n} a_{ii} = a_{11} \cdot a_{22} \cdot ... \cdot a_{nn}$$


**c) Determiınese de forma justificada como alteran el determinante de una
matriz n × n las dos operaciones elementales siguientes:**

- Intercambiar una fila (o columna) por otra fila (o columna).
- Sumar a una fila (o columna) otra fila (o columna) multiplicada por un escalar α.[0.5 puntos]**


Respuesta:
- Intercambiar una fila (o columna) por otra fila (o columna).
El determinante representa el volumen orientado o área, en 2D, definido por los vectores fila de la matriz. Al intercambiar dos filas, se invierte el sentido de ese volumen en el espacio por lo que el signo de su medida (el determinante) se invierte.

Si $A$ es la matriz original y $B$ es la matriz resultante tras intercambiar dos filas, la relación es:

$$\det(B) = -1 \cdot \det(A)$$

Ejemplo:

Sea $A = \begin{bmatrix} 1 & 2 \\ 3 & 4 \end{bmatrix}$.

Su determinante es: $(1 \cdot 4) - (2 \cdot 3) = 4 - 6 = \mathbf{-2}$.

al  intercambiar las filas, obtenemos $B = \begin{bmatrix} 3 & 4 \\ 1 & 2 \end{bmatrix}$.Su determinante es: $(3 \cdot 2) - (4 \cdot 1) = 6 - 4 = \mathbf{2}$.

Resultado: $2 = -1 \cdot (-2)$. El valor cambió de $-2$ a $2$.

---

- Sumar a una fila (o columna) otra fila (o columna) multiplicada por un escalar α.

 Teóricamente, esta operación preserva el volumen orientado (el determinante) de la matriz. Al añadir un múltiplo de una fila a otra, se está aplicando una "transformación"  que desplaza los vectores fila sin cambiar el área o volumen que encierran, por lo que el determinante permanece inalterado.

Si $A$ es la matriz original y $B$ es la matriz resultante tras realizar la operación $R_j \leftarrow R_j + \alpha R_i$, entonces:

$$\det(B) = \det(A)$$


Ejemplo:

Sea $A = \begin{bmatrix} 1 & 2 \\ 3 & 4 \end{bmatrix}$.


$\det(A) = (1 \cdot 4) - (2 \cdot 3) = 4 - 6 = \mathbf{-2}$.


Operación $R_2 \leftarrow R_2 + 2 \cdot R_1$


Fila 2 original ($R_2$): $[3, 4]$

Fila 1 original ($R_1$): $[1, 2]$


Escalar ($\alpha$): $2$

Cálculo: se toma el primer elemento de $R_2$ ($3$) y se suma primer elemento de $R_1$ ($1$) multiplicado por $2$:

$$3 + (2 \cdot 1) = 3 + 2 = \mathbf{5}$$


Tomamos el segundo elemento de $R_2$ ($4$) y le sumamos el segundo elemento de $R_1$ ($2$) multiplicado por $2$:$$4 + (2 \cdot 2) = 4 + 4 = \mathbf{8}$$


Resultado: La nueva Fila 2 es $[5, 8]$.3.


La comprobación del determinanteAhora comparamos la matriz original $A$ con la nueva matriz $B$:

Matriz $A$: $\begin{bmatrix} 1 & 2 \\ 3 & 4 \end{bmatrix} \rightarrow \det(A) = (1 \cdot 4) - (2 \cdot 3) = 4 - 6 = \mathbf{-2}$

Matriz $B$: $\begin{bmatrix} 1 & 2 \\ 5 & 8 \end{bmatrix} \rightarrow \det(B) = (1 \cdot 8) - (2 \cdot 5) = 8 - 10 = \mathbf{-2}$










**d) Investiga sobre el metodo de eliminacion de Gauss con pivoteo parcial e
implementalo para escalonar una matriz (es decir, convertirla en una matriz triangular inferior) a partir de las operaciones elementales descritas en el apartado anterior.[1 punto]**

**Método de Gauss (Eliminación)**
La filosofía de Gauss: "Triangular la Matriz"
Busca que todos los números que están debajo de la diagonal se vuelvan cero:

$$\begin{bmatrix}
Pivote & * & * \\
0 & Pivote & * \\
0 & 0 & Pivote
\end{bmatrix}$$(Donde * es cualquier número y 0 son las zonas que "Eliminadas").

$$A = \begin{bmatrix} 5 & 2 & 9 \\ 1 & 3 & 4 \\ 8 & 6 & 2 \end{bmatrix}$$



Paso 1: Columna 1
-    Pivote Inicial ($A_{11}$): El elemento en la posición diagonal es 5.

-   ¿Es el más grande de su columna? Buscamos en la columna 1: $|5|$, $|1|$, $|8|$. El valor absoluto más grande es 8 (fila 3).


-   Pivoteo Parcial (Intercambio): Intercambiamos la Fila 1 ($R_1$) con la Fila 3 ($R_3$) para garantizar la estabilidad numérica. Contador de intercambios: $1$ (por lo tanto, al final multiplicaremos el resultado por $-1$).
La matriz queda así:

$$A' = \begin{bmatrix} \mathbf{8} & 6 & 2 \\ 1 & 3 & 4 \\ 5 & 2 & 9 \end{bmatrix}$$


-   Eliminación debajo del nuevo pivote (8):  Fila 2 ($R_2$): El factor de eliminación es $1/8 = 0.125$.

$$R_2 \leftarrow R_2 - 0.125 \cdot R_1$$


$$R_2 = [1 - 0.125(8), \quad 3 - 0.125(6), \quad 4 - 0.125(2)] = [0, \quad 2.25, \quad 3.75]$$


Fila 3 ($R_3$): El factor de eliminación es $5/8 = 0.625$.$$R_3 \leftarrow R_3 - 0.625 \cdot R_1$$

$$R_3 = [5 - 0.625(8), \quad 2 - 0.625(6), \quad 9 - 0.625(2)] = [0, \quad -1.75, \quad 7.75]$$

Caluladora:

- 5 - 0.625 * 8 = 0
- 2 - 0.625 * 6 = -1.75
- 9 - 0.625 * 2 = 7.75



La matriz tras esta fase es:$$\begin{bmatrix} 8 & 6 & 2 \\ 0 & \mathbf{2.25} & 3.75 \\ 0 & -1.75 & 7.75 \end{bmatrix}$$


Paso 2: Columna 2

-   Pivote Actual ($A_{22}$): El elemento en la diagonal es 2.25.
-   ¿Es el más grande de su columna hacia abajo? Comparamos $|2.25|$ con $|-1.75|$. Como $2.25$ es el mayor, no hace falta realizar un intercambio.
-   Eliminación debajo de 2.25:
 Fila 3 ($R_3$): El factor es $-1.75 / 2.25 \approx -0.7778$.

 $$R_3 \leftarrow R_3 - (-0.7778) \cdot R_2 \implies R_3 \leftarrow R_3 + 0.7778 \cdot R_2$$

 $$R_3 = [0, \quad -1.75 + 1.75, \quad 7.75 + 0.7778(3.75)] = [0, \quad 0, \quad 10.6667]$$

 La matriz final en su forma Triangular Superior ($U$) es:$$U = \begin{bmatrix} 8 & 6 & 2 \\ 0 & 2.25 & 3.75 \\ 0 & 0 & 10.6667 \end{bmatrix}$$



Cálculo del Determinante

Multiplicamos los elementos de la diagonal principal y aplicamos el cambio de signo por el intercambio de filas realizado en el Paso 1:

$$\det(A) = (-1)^{\text{intercambios}} \cdot (u_{11} \cdot u_{22} \cdot u_{33})$$

$$\det(A) = (-1)^1 \cdot (8 \cdot 2.25 \cdot 10.6667)$$

$$\det(A) = -1 \cdot (18 \cdot 10.6667) = -1 \cdot 192 = -192$$

In [25]:
#--------------------------------------------------------------------------------
# Metodo de eliminacion de Gauss
#  con pivoteo parcial para escalonar una matriz,
#  es decir, convertirla en una matriz triangular inferior)
#--------------------------------------------------------------------------------

# 1. Función para escalonar (Pivoteo parcial + Eliminación)
def escalonar_matriz_pivoteo(matriz):
    A = matriz.astype(float)
    n = A.shape[0]
    intercambios = 0

    for i in range(n):
        # Pivoteo parcial
        max_index = i + np.argmax(np.abs(A[i:, i]))
        if i != max_index:
            A[[i, max_index]] = A[[max_index, i]]
            intercambios += 1

        # Eliminación
        for j in range(i + 1, n):
            if A[i, i] == 0: continue
            factor = A[j, i] / A[i, i]
            A[j, i:] -= factor * A[i, i:]

    return A, intercambios


In [26]:
# Ejemplo de uso:
matriz_2 = np.array([
    [8, 6, 2],
    [1, 3, 4],
    [5, 2, 9]
], dtype=float)


print("\nMatriz  x :             \n", matriz_2)
matriz_escalonada = escalonar_matriz_pivoteo(matriz_2)
# nota: Si la función devuelve (matriz, intercambios) (que es una tupla), y se  asigna a una sola variable
# matriz_escalonada ya no será una matriz, sino una tupla que contiene ambos valores: (matriz_real, numero_intercambios).
# por eso en la linea de abajo se usa "_" . las tuplas ni se pueden modificar
print(f"Matriz escalonada:\n{matriz_escalonada}")


Matriz  x :             
 [[8. 6. 2.]
 [1. 3. 4.]
 [5. 2. 9.]]
Matriz escalonada:
(array([[ 8.  ,  6.  ,  2.  ],
       [ 0.  ,  2.25,  3.75],
       [ 0.  ,  0.  , 10.67]]), 0)


**e) ¿Como se podria calcular el determinante de una matriz haciendo beneficio
de la estrategia anterior y del efecto de aplicar las operaciones elementales pertinentes? Implementa una nueva funcion, determinante gauss, que calcule el determinante de una matriz utilizando eliminacion gaussiana. [0.5 puntos]**

In [27]:
#--------------------------------------------------------------------------------
#
# Determinante de una matriz utilizando eliminacion gaussiana.
#
#--------------------------------------------------------------------------------

# 2. Función para calcular el determinante (usando la anterior)
# Definición RE-ASEGURADA de la función de determinante
def calcular_determinante_gauss(matriz):
    matriz_escalonada, intercambios = escalonar_matriz_pivoteo(matriz)
    det = np.prod(np.diag(matriz_escalonada)) * ((-1) ** intercambios)
    return det

In [28]:
#--------------------------------------------------------------------------------
#
#  Pruebas
#
#--------------------------------------------------------------------------------

# Definimos la matriz
matriz_2 = np.array([
    [8, 6, 2],
    [1, 3, 4],
    [5, 2, 9]
], dtype=float)


resultado_det = calcular_determinante_gauss(matriz_2)
print(f"\nEl determinante calculado por Gauss es: {resultado_det}")


El determinante calculado por Gauss es: 192.0


**f) Obten la complejidad computacional asociada al calculo del determinante
con la definicion recursiva y con el metodo de eliminacion de Gauss con pivoteo parcial.[0.5 puntos]**

### - Definición Recursiva (Expansión por Cofactores / Laplace)
*   **Complejidad:** $O(n!)$ (Factorial).
*   El algoritmo genera una ramificación recursiva de $n$ subproblemas por cada paso, lo que causa un crecimiento explosivo.Es  ineficiente y a partir de $n=10$, el número de operaciones alcanza millones; para matrices mayores, es inviable computacionalmente.

### - Eliminación de Gauss (con Pivoteo Parcial)
*   **Complejidad:** $O(n^3)$ (Cúbica).
*   Se basa en tres ciclos anidados que reducen la matriz a forma triangular. El esfuerzo crece de forma polinómica, no exponencial. Es   Altamente eficiente y capaz de manejar matrices grandes en tiempo real.


**g)  Utilizando numpy.random.rand, genera matrices cuadradas aleatorias de la
forma An ∈ Rn×n, para 2 ≤ n ≤ 10, y confecciona una tabla comparativa del tiempo de ejecucion asociado a cada una de las variantes siguientes, interpretando los resultados:[1 punto]**

$$A_n \in \mathbb{R}^{n \times n}, \text{ para } 2 \leq n \leq 10$$

-  A_n: Nombre de la matriz. La "$n$" es el subíndice que indica que es una matriz cuyo tamaño depende de esa variable $n$

-  ∈ Este símbolo significa "pertenece a". Se usa para decir que un objeto forma parte de un conjunto. La matriz $A_n$ vive dentro del conjunto de los números reales.

-  ℝ^{n×n} :ℝ: Representa el conjunto de los números reales ( decimales, enteros, negativos, positivos, etc.).n×n: Indica las dimensiones de la matriz: $n$ filas por $n$ columnas (una matriz cuadrada). (ℝ^{n×n}): Significa que $A_n$ es una matriz cuadrada de tamaño $n \times n$ llena de números reales.

-  para 2 ≤ n ≤ 10 Define el rango o los límites de la variable $n$. Se  debe generar matrices empezando desde el tamaño $2 \times 2$ hasta el tamaño $10 \times 10$.

Utilizar lo siguiente
-  Utilizando determinante recursivo.
-  Empleando determinante gauss.
-  Haciendo uso de la funcion preprogramada numpy.linalg.det.

In [29]:
#-----------------------------------------------------------
# Generador  de Matrices Aleatorias
#------------------------------------------------------------

def generar_matrices_de_2_a_10():
    matrices = {}
    # Loop from n=2 to n=10
    for n in range(2, 11):
        # Using np.random.rand as requested by the prompt
        matrices[n] = np.random.rand(n, n)
        print(f"Generated matrix of size {n}x{n}")
    return matrices

# Run the generator
todas_las_matrices = generar_matrices_de_2_a_10()

# Iteramos sobre n (de 2 a 10)
for n in range(2, 11):
    # Usamos la variable 'n' dentro de la f-string
    print(f"\nExample: {n}x{n} Matriz generada con np.random.rand:")
    print(todas_las_matrices[n])


Generated matrix of size 2x2
Generated matrix of size 3x3
Generated matrix of size 4x4
Generated matrix of size 5x5
Generated matrix of size 6x6
Generated matrix of size 7x7
Generated matrix of size 8x8
Generated matrix of size 9x9
Generated matrix of size 10x10

Example: 2x2 Matriz generada con np.random.rand:
[[0.03 0.57]
 [0.76 0.88]]

Example: 3x3 Matriz generada con np.random.rand:
[[0.34 0.82 0.11]
 [0.85 0.13 0.4 ]
 [0.8  0.15 0.23]]

Example: 4x4 Matriz generada con np.random.rand:
[[0.72 0.72 0.64 0.69]
 [0.54 0.25 0.35 0.18]
 [0.91 0.58 0.4  0.46]
 [0.95 0.15 0.59 0.51]]

Example: 5x5 Matriz generada con np.random.rand:
[[0.61 0.02 0.87 0.93 0.57]
 [0.7  0.92 0.71 0.15 0.58]
 [0.61 0.42 0.74 0.93 0.93]
 [0.45 0.11 0.98 0.84 0.12]
 [0.92 0.87 0.52 0.59 0.4 ]]

Example: 6x6 Matriz generada con np.random.rand:
[[0.05 0.34 0.8  0.   0.33 0.4 ]
 [0.54 0.92 0.35 0.35 0.74 0.45]
 [0.22 0.45 0.14 0.18 0.5  0.42]
 [0.91 0.36 0.58 0.63 0.01 0.66]
 [0.18 0.96 0.15 0.41 0.09 1.  ]
 [0.5 

In [30]:
 #-------------------------------------------------------------------------------
# Generador  de Matrices Aleatorias & tabla comparativa del tiempo de ejecucion
#-------------------------------------------------------------------------------
from functools import lru_cache
np.random.seed(42)

# 2. Generar Tabla de Resultados de Determinantes
print("Tabla de Resultados de Determinantes ")
print("| Matriz (N x N) | Det (Laplace)| Det (Gauss)   | Det (Numpy)   |")
print("|----------------|--------------|---------------|---------------|")
from functools import lru_cache
np.random.seed(42)
max = 11
for n in range(2, max):
    matriz = np.random.rand(n, n)
    det_a = determinante_recursivo_laplace(matriz)
    det_g = calcular_determinante_gauss(matriz)
    det_n = np.linalg.det(matriz)
    print(f"| {n:2d}             | {det_a:+.4f}      | {det_g:+.4f}       | {det_n:+.4f}       |")

# --- COMPARACIÓN DE TIEMPOS ---
print()
print("Tabla de Resultados de Comparacion de tiempos de ejecucion  ")
print(f"{'Tamaño (n)':<12} | {'Laplace (s)':<12} | {'Gauss (s)':<12} | {'NumPy (s)':<12}")
print("-" * 55)

for n in range(3, 11):
    A = np.random.rand(n, n)

    # Laplace
    start = time.perf_counter()
    determinante_recursivo_laplace(A)
    t_laplace = time.perf_counter() - start

    # Gauss
    start = time.perf_counter()
    calcular_determinante_gauss(A)
    t_gauss = time.perf_counter() - start

    # NumPy
    start = time.perf_counter()
    np.linalg.det(A)
    t_numpy = time.perf_counter() - start

    print(f"{n:<12} | {t_laplace:<12.6f} | {t_gauss:<12.6f} | {t_numpy:<12.6f}")

Tabla de Resultados de Determinantes 
| Matriz (N x N) | Det (Laplace)| Det (Gauss)   | Det (Numpy)   |
|----------------|--------------|---------------|---------------|
|  2             | -0.4717      | -0.4717       | -0.4717       |
|  3             | -0.0912      | -0.0912       | -0.0912       |
|  4             | -0.0013      | -0.0013       | -0.0013       |
|  5             | +0.0603      | +0.0603       | +0.0603       |
|  6             | +0.0793      | +0.0793       | +0.0793       |
|  7             | -0.0173      | -0.0173       | -0.0173       |
|  8             | -0.0426      | -0.0426       | -0.0426       |
|  9             | -0.0182      | -0.0182       | -0.0182       |
| 10             | +0.0385      | +0.0385       | +0.0385       |

Tabla de Resultados de Comparacion de tiempos de ejecucion  
Tamaño (n)   | Laplace (s)  | Gauss (s)    | NumPy (s)   
-------------------------------------------------------
3            | 0.000071     | 0.000131     | 0.000031    
4 

Análisis Comparativo de Métodos para el Cálculo de Determinantes

Para el cálculo del determinante de matrices $A_n \in \mathbb{R}^{n \times n}$, se analizaron tres enfoques distintos:

1. Método Recursivo (Laplace): Implementa una expansión por cofactores. Su complejidad es $O(n!)$, lo que implica un crecimiento factorial. Es el método más ineficiente. Incluso para matrices pequeñas ($n > 10$), el tiempo de ejecución aumenta drásticamente, haciendo que sea inviable para aplicaciones de ciencia de datos.

2. Método de Eliminación Gaussiana: Reduce la matriz a una forma triangular superior mediante eliminación. Su complejidad es $O(n^3)$. Es mucho más eficiente que la recursión. Para un $n$ moderado, es la mejor opción.

3. Función Preprogramada (numpy.linalg.det): Utiliza algoritmos de álgebra lineal altamente optimizados. Es el método estándar mas utilizzado.  


# **Ejercicio 2: El Método del Descenso de Gradiente**
Este notebook contiene la resolución detallada y paso a paso del ejercicio 2 de la asignatura *Matemáticas para la IA*. Analizaremos la optimización mediante el descenso de gradiente tanto en funciones de una variable como de varias variables, contrastando los resultados experimentales con el análisis analítico.

## **Apartado a) Implementación general del Descenso de Gradiente**

El algoritmo del descenso de gradiente es un método iterativo de optimización de primer orden utilizado para encontrar mínimos locales de una función. La regla de actualización para un vector de parámetros $x$ en la iteración $k+1$ viene dada por:

$$x_{k+1} = x_k - \gamma \nabla f(x_k)$$

Donde $\gamma$ es el ratio de aprendizaje (*learning rate*) y $\nabla f(x_k)$ es el gradiente de la función en el punto actual. El algoritmo se detiene cuando la norma euclídea del gradiente es menor que una tolerancia predefinida ($||\nabla f(x)||_2 < \text{tol}$) o cuando se alcanza el número máximo de iteraciones (`maxit`).

In [31]:
import numpy as np

def descenso_gradiente(grad_f, x0, gamma, tol=1e-12, maxit=1e5):
    """
    Implementación del algoritmo de descenso de gradiente para n variables.
    """
    # Convertimos el punto inicial a un array de numpy de tipo flotante
    x = np.array(x0, dtype=float)

    for i in range(int(maxit)):
        grad = grad_f(x)
        norma_grad = np.linalg.norm(grad, ord=2)

        # Criterio de parada
        if norma_grad < tol:
            print(f"Convergencia alcanzada en la iteración {i}. Norma del gradiente: {norma_grad:.2e}")
            return x

        # Actualización
        x = x - gamma * grad

    print(f"Se alcanzó el número máximo de iteraciones ({maxit}) sin cumplir la tolerancia.")
    return x

---
## **Apartado b) Estudio de la función de una variable** $f(x) = 3x^4 + 4x^3 - 12x^2 + 7$

### **Estudio Analítico Preliminar**
1. **Primera derivada (Gradiente):**
   $$f'(x) = 12x^3 + 12x^2 - 24x = 12x(x^2 + x - 2) = 12x(x + 2)(x - 1)$$
   Puntos críticos en $x = 0$, $x = 1$ y $x = -2$.

2. **Segunda derivada (Clasificación):**
   $$f''(x) = 36x^2 + 24x - 24$$
   * $x = 0 \implies f''(0) = -24 < 0 \implies$ **Máximo local** en $(0, 7)$.
   * $x = 1 \implies f''(1) = 36 > 0 \implies$ **Mínimo local** en $(1, 2)$.
   * $x = -2 \implies f''(-2) = 72 > 0 \implies$ **Mínimo global** en $(-2, -25)$.

In [32]:
# Definimos el gradiente de f(x)
def grad_f(x):
    return np.array([12*x[0]**3 + 12*x[0]**2 - 24*x[0]])

tol_b = 1e-12
maxit_b = 1e5
x0_b = [3.0]

print("--- EXPERIMENTO I (gamma = 0.001) ---")
sol_I = descenso_gradiente(grad_f, x0=x0_b, gamma=0.001, tol=tol_b, maxit=maxit_b)
print(f"Resultado: x = {sol_I[0]:.6f}\n")

print("--- EXPERIMENTO II (gamma = 0.01) ---")
sol_II = descenso_gradiente(grad_f, x0=x0_b, gamma=0.01, tol=tol_b, maxit=maxit_b)
print(f"Resultado: x = {sol_II[0]:.6f}\n")

print("--- EXPERIMENTO IV (gamma = 0.1) ---")
try:
    sol_IV = descenso_gradiente(grad_f, x0=x0_b, gamma=0.1, tol=tol_b, maxit=maxit_b)
    print(f"Resultado: x = {sol_IV[0]}\n")
except Exception as e:
    print(f"Error detectado por divergencia: {e}\n")

print("--- EXPERIMENTO V (x0 = 0) ---")
sol_V = descenso_gradiente(grad_f, x0=[0.0], gamma=0.001, tol=tol_b, maxit=maxit_b)
print(f"Resultado: x = {sol_V[0]:.6f}\n")

--- EXPERIMENTO I (gamma = 0.001) ---
Convergencia alcanzada en la iteración 831. Norma del gradiente: 9.91e-13
Resultado: x = 1.000000

--- EXPERIMENTO II (gamma = 0.01) ---
Convergencia alcanzada en la iteración 31. Norma del gradiente: 8.38e-13
Resultado: x = -2.000000

--- EXPERIMENTO IV (gamma = 0.1) ---


/tmp/ipykernel_13954/945039322.py:3: RuntimeWarning: overflow encountered in scalar power
  return np.array([12*x[0]**3 + 12*x[0]**2 - 24*x[0]])
/tmp/ipykernel_13954/945039322.py:3: RuntimeWarning: invalid value encountered in scalar subtract
  return np.array([12*x[0]**3 + 12*x[0]**2 - 24*x[0]])


Se alcanzó el número máximo de iteraciones (100000.0) sin cumplir la tolerancia.
Resultado: x = nan

--- EXPERIMENTO V (x0 = 0) ---
Convergencia alcanzada en la iteración 0. Norma del gradiente: 0.00e+00
Resultado: x = 0.000000



### **Interpretación de los Resultados de la Función de Una Variable**

* **III) Contraste Experimentos I y II:**
  Con $\gamma = 0.001$, los pasos son cortos y el algoritmo cae de forma segura en el mínimo local más cercano a su punto de partida ($x \approx 1.0$). Al aumentar el ratio a $\gamma = 0.01$, el primer paso es tan grande que "salta" el bache del mínimo local y del máximo, aterrizando directamente en la cuenca de atracción del mínimo global ($x \approx -2.0$). El *learning rate* altera drásticamente el destino en superficies no convexas.

* **IV) Interpretación del Experimento IV ($\gamma = 0.1$):**
  Un ratio excesivamente alto provoca que el tamaño del paso sea mayor que el ancho de la propia cuenca de optimización. El algoritmo empieza a **oscilar de manera divergente**, alejándose cada vez más del origen en cada iteración hasta sufrir un desbordamiento numérico (*Overflow/NaN*).

* **V) Interpretación del Experimento V ($x_0 = 0$):**
  **No es un resultado deseable.** El algoritmo se queda atrapado instantáneamente en un **máximo local**. Esto ocurre porque en cualquier punto crítico el gradiente es cero ($f'(0) = 0$), lo que anula el paso de actualización ($x_{k+1} = x_k - \gamma \cdot 0$). El descenso de gradiente clásico es incapaz de distinguir un máximo de un mínimo si empieza exactamente sobre él.

---
## **Apartado c) Estudio de la función de varias variables** $g(x,y) = x^2 + y^3 + 3xy + 1$

### **Estudio Analítico Preliminar**
1. **Gradiente ($\nabla g$):**
   $$\nabla g(x,y) = \begin{pmatrix} 2x + 3y \\ 3y^2 + 3x \end{pmatrix} = \mathbf{0}$$
   Puntos críticos en **$P_1 = (0,0)$** y **$P_2 = (-2.25, 1.5)$**.

2. **Matriz Hessiana ($H_g$):**
   $$H_g(x,y) = \begin{pmatrix} 2 & 3 \\ 3 & 6y \end{pmatrix}$$
   * **Para $P_1(0,0)$:** $\det(H_g) = -9 < 0 \implies$ **Punto de Silla (Saddle Point)**.
   * **Para $P_2(-2.25, 1.5)$:** $\det(H_g) = 9 > 0$ y $2 > 0 \implies$ **Mínimo Local**.

In [33]:
# Definimos el gradiente de g(x, y)
def grad_g(v):
    x, y = v[0], v[1]
    return np.array([2*x + 3*y, 3*y**2 + 3*x])

gamma_c = 0.01
tol_c = 1e-12
maxit_c = 1e5

print("--- EXPERIMENTO 1 (Punto inicial [-1, 1]) ---")
sol_c_I = descenso_gradiente(grad_g, x0=[-1.0, 1.0], gamma=gamma_c, tol=tol_c, maxit=maxit_c)
print(f"Resultado: P = ({sol_c_I[0]:.4f}, {sol_c_I[1]:.4f})\n")

print("--- EXPERIMENTO 2 (Punto inicial [0, 0]) ---")
sol_c_II = descenso_gradiente(grad_g, x0=[0.0, 0.0], gamma=gamma_c, tol=tol_c, maxit=maxit_c)
print(f"Resultado: P = ({sol_c_II[0]:.4f}, {sol_c_II[1]:.4f})")

--- EXPERIMENTO 1 (Punto inicial [-1, 1]) ---
Convergencia alcanzada en la iteración 3139. Norma del gradiente: 9.99e-13
Resultado: P = (-2.2500, 1.5000)

--- EXPERIMENTO 2 (Punto inicial [0, 0]) ---
Convergencia alcanzada en la iteración 0. Norma del gradiente: 0.00e+00
Resultado: P = (0.0000, 0.0000)


### **III) Explicación y Contraste de Resultados en $g(x,y)$**

* **Desde $x_0 = (-1,1)$:** El algoritmo converge con éxito al punto $(-2.25, 1.50)$. Esto valida experimentalmente el análisis matemático, ya que este punto demostró ser un **mínimo local** real (estable).

* **Desde $x_0 = (0,0)$:** El algoritmo se congela en la iteración 0. **No es un resultado deseable** porque $(0,0)$ es un **punto de silla**. En este tipo de puntos, la función tiene pendientes de bajada en ciertas direcciones, pero como el gradiente en la coordenada exacta es estrictamente nulo ($\nabla g = 0$), el algoritmo clásico se detiene asumiendo erróneamente que ha terminado con éxito.